### PR TITLE
SpecialWave added

### PR DESCRIPTION
**Pull request by Cataract92**
Refactored Specials
Added 5. SpecialWave

### CODE REVIEW

File: wurst/Creeps/UnitSpawnSystem.wurst [hunk 1]
```diff
@@ -21,12 +26,11 @@ int level = 1
 	
 		
 function onSpawnTime()	
-	if level mod 7 == 0 and specialWaves.getSize() > 0
```
Comment (Frotty) [inline wurst/Creeps/UnitSpawnSystem.wurst:24]: please re-add these checks to avoid nullpointer errors

File: wurst/Creeps/specials/Special.wurst [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package Special
+
+import Creep
+import ObjEditingNatives
+
+//COMPILETIME
```
Comment (Frotty) [inline wurst/Creeps/specials/Special.wurst]: Use the Human Readable classes please (UnitDefinition/AbilityDefinition)

File: wurst/_abstract/buff/Buff.wurst [hunk 3]
```diff
@@ -63,35 +63,44 @@ init
...
+					// Correct attack, apply Buffs
+					if victimData != null and victimData.hasBuffs()
+						for bff in victimData.getBuffs()
+							if not bff.done
+								bff.defenseModifier()
+					if attackerData != null and attackerData.hasBuffs()
 						for bff in attackerData.getBuffs()
 							if not bff.done
 								bff.attackModifier()
-					modifyDamage(0)
```
Comment (Frotty) [inline wurst/_abstract/buff/Buff.wurst:94]: Remove this and give the SPecial a Buff extended NormalShieldBuff instead. use defenseModifier as onDamage

File: wurst/_abstract/entity/Entity.wurst [hunk 4]
```diff
@@ -130,6 +131,8 @@ abstract public class UnitEntity extends Entity
 	function onDeath(unit killer)
 		done = true
 			
+	function onDamage(unit attacker)
```
Comment (Frotty) [inline wurst/_abstract/entity/Entity.wurst]: same for these modifications

File: wurst/Creeps/specials/Special.wurst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	u.setString("fnam", "*Slow")
+	u.setString("fnsf", "")
+	u.setString("ftip", "Slow")
+	u.setString("fube", "This Unit is slowed.")
+	u.setString("ftat", "")
+	u.setString("fart", "ReplaceableTextures\\CommandButtons\\BTNSlow.blp")
+//
+
+public class Special extends Creep
+	effect glow
```
Comment (Frotty) [inline wurst/Creeps/specials/Special.wurst:95]: make class abstract

File: wurst/Creeps/specials/SpecialFive.wurst [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package SpecialFive
+
+import Creep
+import PresetBuffs
+import Special
+
+class SlowDebuff extends NormalBuff
```
Comment (Frotty) [inline wurst/Creeps/specials/SpecialFive.wurst]: extend the NormalShieldBuff and then the defense modifier for onDamage

File: wurst/Creeps/specials/SpecialFive.wurst [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	construct(UnitEntity victim, real dur, int abilId)
+		super(victim, dur, abilId, true)
+
+	override function defenseModifier()
+		super.defenseModifier()
```
Comment (Frotty) [inline wurst/Creeps/specials/SpecialFive.wurst]: These super calls are empty, please remove

File: wurst/Creeps/specials/SpecialFive.wurst [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+class Spawn extends Special
+	construct(vec3 pos, int typId, int bounty, vec2 target)
+		super(pos, typId, bounty, target)
+		
+	override function onDamage(unit attacker)
+		new SlowDebuff(this,2.,'A022')
+
+public class SpecialFive extends Special
+	
+	static constant real SPAWNTIME = 20
```
Comment (Frotty) [inline wurst/Creeps/specials/SpecialFive.wurst]: you can leave out types on constants

File: wurst/Creeps/specials/SpecialTwo.wurst [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import Special
+
+public class SpecialTwo extends Special
+
+	construct(vec3 pos, int typId, int bounty, vec2 target)
+		super(pos, typId, bounty, target)
+
+	override function onDeath(unit killer)
+		var ppos = getPos()
+		if actor.getTypeId() == 'u00B'
```
Comment (Frotty) [inline wurst/Creeps/specials/SpecialTwo.wurst]: Put ids into constants please

File: wurst/_abstract/entity/Entity.wurst [hunk 10]
```diff
@@ -75,6 +75,7 @@ public abstract class Entity
 		
 		let startZVelocity = ((-gravity.z * t) / 2 - pos.z/t + e/t)
 		this.setVel( vec3( Cos(tangle.radians()) * speed, Sin(tangle.radians()) * speed, startZVelocity) )
+		real i = ANIMATION_PERIOD + 0
```
Comment (Frotty) [inline wurst/_abstract/entity/Entity.wurst]: ??

File: wurst/Creeps/specials/SpecialFive.wurst [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+		super(pos, typId, bounty, target)
+
+	override function update()
+		super.update()
+		active = true
+		spawntimer += ANIMATION_PERIOD
+		if spawntimer >= SPAWNTIME
+			spawntimer = 0
+			for i = 1 to SPAWNAMOUNT
+				new Spawn(actor.getPos3(),SPAWNID,0,middle)
```
Comment (Frotty) [inline wurst/Creeps/specials/SpecialFive.wurst]: Spawning multiple units at once looks ugly and might block the caster, instead let them come after each other. Maybe keep the Special in some channel/cast animation while it spawns the minions.
Maybe also a sound.

File: wurst/Creeps/specials/SpecialFive.wurst [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+	override function defenseModifier()
+		super.defenseModifier()
+			
+	override function attackModifier()
+		super.attackModifier()
+	
+
+class Spawn extends Special
+	construct(vec3 pos, int typId, int bounty, vec2 target)
+		super(pos, typId, bounty, target)
```
Comment (Frotty) [inline wurst/Creeps/specials/SpecialFive.wurst:28]: Make the appearance more fancy, play spawn animation, some special effect

File: wurst/_abstract/entity/Entity.wurst (hunk 10 above)
Comment (Cataract92) [inline wurst/_abstract/entity/Entity.wurst]: This was just a test I forgot to remove completly

File: wurst/Creeps/UnitSpawnSystem.wurst [hunk 13]
```diff
@@ -24,9 +29,8 @@ function onSpawnTime()
 	if level mod 7 == 0 and specialWaves.getSize() > 0
 		specialWaves.dequeue().spawn()
 	else
-		if normalWaves.getSize() > 0
-			normalWaves.dequeue().spawn()
-		if level mod 4 == 0 and normalWaves.getSize() > 0
+		normalWaves.dequeue().spawn()
```
Comment (Frotty) [inline wurst/Creeps/UnitSpawnSystem.wurst]: check if it has members before dequeuing to avoid errors at the end of the game.

**Comment by Frotty** [comment]
nice thx

**Pull request merged**